hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -915,6 +917,11 @@ private void finishActiveMasterInitialization(MonitoredTask status)
       this.tableDescriptors.getAll();
     }
 
+    // check cluster Id stored in ZNode before, and use it to indicate if a cluster has been
+    // restarted with an existing Zookeeper quorum.
+    isClusterRestartWithExistingZNodes =
```

Comment:
Hm... This check alone is not enough I fear..  Say there is a cluster recreate over existing data and so no zk node for clusterId.  We will get true here.  In next lines, we will create and write the zk node.   Assume immediately after that, the HM restarted.  Now after restart, when we come here, we have zk data for clusterId, and we will end up deleting the Meta dir !!
After cluster recreate we will come to know this boolean value before creation of zk node.  So we need to record that info some how somewhere so that even if there is a restart before we reach to the Meta bootstrap.  Even that info should be there in InitMetaProcedure also.. If there is an HM restart after the submit of this proc, the next time we know the boolean status.
Comments?

Reply:
Thanks Anoop, if I get your suggestion correctly, should we add a master procedure (or a state in InitMetaProcedure) which record this boolean? this will have the assumption with the WAL exists when cluster recreates.

Reply:
Ya at proc level..  But before that itself one more thing.
Say there is a cluster recreate over existing data and so no zk node for clusterId. We will get true for 'isClusterRestartWithExistingZNodes '. In next lines, we will create and write the zk node for clusterId.  Now assume after executing that lines, the HM restarted.  So the zk node was created but the InitMetaProc was NOT submitted.  Now after restart, when we come here, we have zk data for clusterId.  So 'isClusterRestartWithExistingZNodes ' will become false.  Now this time the InitMetaProc started and as part of that we will end up deleting the Meta dir.
So this says the need to keep this boolean info somewhere once we find that and even before creating the zk node for ClusterId.  Am I making the concern clear this time?

Reply:
> So this says the need to keep this boolean info somewhere once we find that and even before creating the zk node for ClusterId. Am I making the concern clear this time?

ack and I got the concern. 

but before I move to implement the proc level information, let's wait a bit on @Apache9 for the question on META's `tableInfo` and `partial meta`.